# Hamoye Internship Stage A Codes

In [46]:
import numpy as np
import pandas as pd

In [47]:
fuel_data = pd.read_csv('fuel_ferc1.csv')

In [48]:
fuel_data

record_id  utility_id_ferc1  report_year  \
0        f1_fuel_1994_12_1_0_7                 1         1994   
1       f1_fuel_1994_12_1_0_10                 1         1994   
2        f1_fuel_1994_12_2_0_1                 2         1994   
3        f1_fuel_1994_12_2_0_7                 2         1994   
4       f1_fuel_1994_12_2_0_10                 2         1994   
...                        ...               ...          ...   
29518  f1_fuel_2018_12_12_0_13                12         2018   
29519   f1_fuel_2018_12_12_1_1                12         2018   
29520  f1_fuel_2018_12_12_1_10                12         2018   
29521  f1_fuel_2018_12_12_1_13                12         2018   
29522  f1_fuel_2018_12_12_1_14                12         2018   

           plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
0                  rockport                coal       ton        5377489.0   
1      rockport total plant                coal       ton       10486945.0   
2                    gorgas                coal       ton        2978683.0   
3                     barry                coal       ton        3739484.0   
4                 chickasaw                 gas       mcf          40533.0   
...                     ...                 ...       ...              ...   
29518    neil simpson ct #1                 gas       mcf          18799.0   
29519  cheyenne prairie 58%                 gas       mcf         806730.0   
29520     lange ct facility                 gas       mcf         104554.0   
29521       wygen 3 bhp 52%                coal       ton         315945.0   
29522       wygen 3 bhp 52%                 gas       mcf          17853.0   

       fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
0                   16.590                      18.59   
1                   16.592                      18.58   
2                   24.130                      39.72   
3                   23.950                      47.21   
4                    1.000                       2.77   
...                    ...                        ...   
29518                1.059                       4.78   
29519                1.050                       3.65   
29520                1.060                       4.77   
29521               16.108                       3.06   
29522                1.059                       0.00   

       fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
0                             18.53                1.121  
1                             18.53                1.120  
2                             38.12                1.650  
3                             45.99                1.970  
4                              2.77                2.570  
...                             ...                  ...  
29518                          4.78                9.030  
29519                          3.65                6.950  
29520                          4.77                8.990  
29521                         14.76                1.110  
29522                          0.00               11.680  

[29523 rows x 11 columns]

## Question 3

In [49]:
fuel_data.groupby('fuel_type_code_pudl')['fuel_cost_per_unit_burned'].mean()

fuel_type_code_pudl
coal         67.421830
gas          13.659397
nuclear    4955.157002
oil         168.877086
other        18.253856
waste        19.518122
Name: fuel_cost_per_unit_burned, dtype: float64

gas has the lowest average 'fuel_cost_per_unit_burned' with a value of 13.659397


## Question 4

In [50]:
fuel_data["fuel_mmbtu_per_unit"].describe().round(2)

count    29523.00
mean         8.49
std         10.60
min          0.00
25%          1.02
50%          5.76
75%         17.01
max        341.26
Name: fuel_mmbtu_per_unit, dtype: float64

Standard Deviation is 10.60 and 75th percentile is 17.01(to 2 decimal places)

## Question 5

In [51]:
skewness = fuel_data['fuel_qty_burned'].skew().round(2)
skewness 

15.85

In [52]:
kurtosis = fuel_data['fuel_qty_burned'].kurtosis().round(2)
kurtosis

651.37

## Question 6

In [53]:
fuel_data.isnull().sum()

record_id                         0
utility_id_ferc1                  0
report_year                       0
plant_name_ferc1                  0
fuel_type_code_pudl               0
fuel_unit                       180
fuel_qty_burned                   0
fuel_mmbtu_per_unit               0
fuel_cost_per_unit_burned         0
fuel_cost_per_unit_delivered      0
fuel_cost_per_mmbtu               0
dtype: int64

In [54]:
total = fuel_data["fuel_unit"].count()
total

29343

In [55]:
missing = fuel_data["fuel_unit"].isnull().sum()
missing

180

In [56]:
per = (missing/total) * 100
per

0.6134342091810653

In [57]:
per.round(3)

0.613

## Question 7

In [58]:
fuel_data['fuel_unit'].dtype

dtype('O')

In [59]:
fuel_data.groupby( 'fuel_unit' )[ 'fuel_unit' ].count()

fuel_unit
bbl        7998
gal          84
gramsU      464
kgU         110
mcf       11354
mmbtu       180
mwdth        95
mwhth       100
ton        8958
Name: fuel_unit, dtype: int64

In [60]:
fuel_data['fuel_unit'].mode()

0    mcf
dtype: object

Categorical and mode

## Question 8

In [90]:
data_corr = fuel_data.corr()
data_corr

utility_id_ferc1  report_year  fuel_qty_burned  \
utility_id_ferc1                      1.000000     0.093323        -0.057447   
report_year                           0.093323     1.000000         0.012952   
fuel_qty_burned                      -0.057447     0.012952         1.000000   
fuel_mmbtu_per_unit                  -0.066946    -0.110853        -0.080946   
fuel_cost_per_unit_burned            -0.037863     0.013599        -0.018535   
fuel_cost_per_unit_delivered         -0.016414    -0.014043        -0.003551   
fuel_cost_per_mmbtu                   0.006122     0.010261        -0.001896   

                              fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
utility_id_ferc1                        -0.066946                  -0.037863   
report_year                             -0.110853                   0.013599   
fuel_qty_burned                         -0.080946                  -0.018535   
fuel_mmbtu_per_unit                      1.000000                  -0.010034   
fuel_cost_per_unit_burned               -0.010034                   1.000000   
fuel_cost_per_unit_delivered            -0.009039                   0.011007   
fuel_cost_per_mmbtu                     -0.005884                  -0.000437   

                              fuel_cost_per_unit_delivered  \
utility_id_ferc1                                 -0.016414   
report_year                                      -0.014043   
fuel_qty_burned                                  -0.003551   
fuel_mmbtu_per_unit                              -0.009039   
fuel_cost_per_unit_burned                         0.011007   
fuel_cost_per_unit_delivered                      1.000000   
fuel_cost_per_mmbtu                              -0.000109   

                              fuel_cost_per_mmbtu  
utility_id_ferc1                         0.006122  
report_year                              0.010261  
fuel_qty_burned                         -0.001896  
fuel_mmbtu_per_unit                     -0.005884  
fuel_cost_per_unit_burned               -0.000437  
fuel_cost_per_unit_delivered            -0.000109  
fuel_cost_per_mmbtu                      1.000000

In [91]:
fcpub = data_corr["fuel_cost_per_unit_burned"]
fcpub

utility_id_ferc1               -0.037863
report_year                     0.013599
fuel_qty_burned                -0.018535
fuel_mmbtu_per_unit            -0.010034
fuel_cost_per_unit_burned       1.000000
fuel_cost_per_unit_delivered    0.011007
fuel_cost_per_mmbtu            -0.000437
Name: fuel_cost_per_unit_burned, dtype: float64

- Second lowest value is -0.018535 which corresponds with fuel_qty_burned 
- Third lowest value is -0.010034 which corresponds with fuel_mmbtu_per_unit

## Question 9

In [62]:
per_data = fuel_data.groupby(['fuel_type_code_pudl','report_year'])['fuel_cost_per_unit_burned'].sum()
per_data

fuel_type_code_pudl  report_year
coal                 1994           14984.572
                     1995           14571.785
                     1996           12694.803
                     1997           43742.178
                     1998           11902.597
                                      ...    
waste                2014             591.594
                     2015             629.270
                     2016             539.079
                     2017             452.079
                     2018             660.972
Name: fuel_cost_per_unit_burned, Length: 148, dtype: float64

In [63]:
coal_1994 = per_data.loc[('coal',1994)]
coal_1998 = per_data.loc[('coal',1998)]
per_change = ((coal_1998 - coal_1994)/coal_1994) * 100

In [64]:
per_change

-20.56765451826049

In [65]:
per_change.round()

-21.0

## Question 10

In [66]:
avg = fuel_data.groupby("report_year")["fuel_cost_per_unit_delivered"].mean()
avg

report_year
1994       63.636060
1995       32.735269
1996     9196.705948
1997    11140.197239
1998      287.154420
1999       25.551627
2000      985.362877
2001       60.050396
2002       47.594361
2003       55.663493
2004      139.524275
2005       41.438184
2006       38.657484
2007       43.325023
2008       58.588197
2009      652.694163
2010       91.862105
2011       59.774667
2012       60.994502
2013      172.307591
2014      192.737183
2015      326.535511
2016      103.901761
2017       46.196861
2018      499.269966
Name: fuel_cost_per_unit_delivered, dtype: float64

In [67]:
avg.max()

11140.19723948813

In [68]:
avg.argmax()

3

The year with the highest average fuel_cost_per_unit_delivered is 1997.